EventFilter/EcalRawToDigi/python/ecalDigis_cff.py

```diff
@@ -15,29 +16,33 @@
 # process modifier to run on GPUs
 from Configuration.ProcessModifiers.gpu_cff import gpu
 
-# ECAL conditions used by the unpacker running on GPU
-from EventFilter.EcalRawToDigi.ecalElectronicsMappingGPUESProducer_cfi import ecalElectronicsMappingGPUESProducer
+from Configuration.StandardSequences.Accelerators_cff import *
+from HeterogeneousCore.AlpakaCore.ProcessAcceleratorAlpaka_cfi import ProcessAcceleratorAlpaka
+
+# ECAL conditions used by the portable unpacker
+from EventFilter.EcalRawToDigi.ecalElectronicsMappingHostESProducer_cfi import ecalElectronicsMappingHostESProducer
+
+# alpaka ECAL unpacker
+from EventFilter.EcalRawToDigi.ecalRawToDigiPortable_cfi import ecalRawToDigiPortable as _ecalRawToDigiPortable
+ecalDigisPortable = _ecalRawToDigiPortable.clone()
 
-# ECAL unpacker running on GPU
-from EventFilter.EcalRawToDigi.ecalRawToDigiGPU_cfi import ecalRawToDigiGPU as _ecalRawToDigiGPU
-ecalDigisGPU = _ecalRawToDigiGPU.clone()
+# extend the SwitchProducer to add a case to copy the ECAL digis from the accelerator to CPU and covert them from SoA to legacy format
+from EventFilter.EcalRawToDigi.ecalDigisFromPortableProducer_cfi import ecalDigisFromPortableProducer as _ecalDigisFromPortableProducer
 
-# extend the SwitchProducer to add a case to copy the ECAL digis from GPU to CPU and covert them from SoA to legacy format
-from EventFilter.EcalRawToDigi.ecalCPUDigisProducer_cfi import ecalCPUDigisProducer as _ecalCPUDigisProducer
 gpu.toModify(ecalDigis,
-    # copy the ECAL digis from GPU to CPU and covert them from SoA to legacy format
-    cuda = _ecalCPUDigisProducer.clone(
-        digisInLabelEB = ('ecalDigisGPU', 'ebDigis'),
-        digisInLabelEE = ('ecalDigisGPU', 'eeDigis'),
+    # copy the ECAL digis from the accelerator to CPU and covert them from SoA to legacy format
+    cuda = _ecalDigisFromPortableProducer.clone(
+        digisInLabelEB = 'ecalDigisPortable:ebDigis',
+        digisInLabelEE = 'ecalDigisPortable:eeDigis',
         produceDummyIntegrityCollections = True
```

Comment:
Long term the `SwitchProducerCUDA` here and for the uncalibrated rechits producer should be replaced by only the alpaka module. For the unpacker the legacy version needs to run as well at the moment because some functionality is not present in the CUDA/alpaka ones.

I was not aware of the `alpaka` modifier. I think this is a good idea to use it here.

Reply:
If I'm not mistaken the two branches (`cpu` and `cuda`) are identical.
Would it work to keep only the `cpu` branch ?

Reply:
`SwitchProducer` should raise an exception if `cuda` case is chosen, but only `cpu` case is specified.

One way to reduce code duplication could be along
```py
_ecalDigisPortable = _ecalDigisFromPortableProducer.clone(
    digisInLabelEB = 'ecalDigisPortable:ebDigis',
    digisInLabelEE = 'ecalDigisPortable:eeDigis',
    produceDummyIntegrityCollections = True
)
alpaka.toModify(ecalDigis, cpu=_ecalDigisPortable.clone(), cuda=_ecalDigisPortable.clone())
```

Reply:
Actually, no, that's not how `SwitchProducer` works: if there is no `cuda` branch, it simply uses the `cpu` branch unconditionally.

We use this feature regularly in the offline code (`whatever.cpu` implements the CPU-only functionality, and `whatver.cuda` is added by the `gpu` modifier)

You can check it with a simple configuration:
```python
import FWCore.ParameterSet.Config as cms

from HeterogeneousCore.CUDACore.SwitchProducerCUDA import SwitchProducerCUDA
from HeterogeneousCore.CUDACore.ProcessAcceleratorCUDA import ProcessAcceleratorCUDA

process = cms.Process( "HLT" )
process.options.wantSummary = True
process.maxEvents.input = 100

process.ProcessAcceleratorCUDA = ProcessAcceleratorCUDA()

process.source = cms.Source( "EmptySource" )

process.statusOnGPU = SwitchProducerCUDA(
    cpu = cms.EDProducer( "BooleanProducer",
        value = cms.bool( False )
    ),
    #cuda = cms.EDProducer( "BooleanProducer",
    #    value = cms.bool( True )
    #),
)

process.statusOnGPUFilter = cms.EDFilter( "BooleanFilter",
    src = cms.InputTag( "statusOnGPU" )
)

process.Status_OnCPU = cms.Path( process.statusOnGPU + ~process.statusOnGPUFilter )
process.Status_OnGPU = cms.Path( process.statusOnGPU + process.statusOnGPUFilter )
```

If you uncomment the `process.statusOnGPU.cuda` branch, the `statusOnGPU` will produce `True` or `False` depending if there is a CUDA GPU or not.

If you keep it commented, it will always take the `cpu` branch and always produce `False`.

Reply:
> if there is no `cuda` branch, it simply uses the `cpu` branch unconditionally.

Thanks, I stand corrected (got confused).

Reply:
The possible `cuda` branch could be removed with `alpaka.toModify(ecalDigis, cuda=None)`, that seems to work also when the `cuda` branch is not present. So one option could be
```py
alpaka.toModify(ecalDigis,
    cpu = _ecalDigisFromPortableProducer.clone(...),
    cuda = None
)
```

Reply:
Trying this option I get
```
----- Begin Fatal Exception 16-Jan-2024 18:01:24 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named step3_RAW2DIGI_RECO_VALIDATION_DQM.py
Exception Message:
 unknown python problem occurred.
KeyError: 'Unknown parameter name cuda specified while calling Modifier'

At:
  /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_0_pre2/src/FWCore/ParameterSet/python/Config.py(1773): _raiseUnknownKey
  /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_0_pre2/src/FWCore/ParameterSet/python/Mixins.py(755): _modifyParametersFromDict
  /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_0_pre2/src/FWCore/ParameterSet/python/Config.py(1764): __call__
  /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_0_pre2/src/FWCore/ParameterSet/python/Config.py(1870): _toModify
  /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_0_pre2/src/FWCore/ParameterSet/python/Config.py(1862): toModify
  /data/user/treis/ecal/run3/dev/alpaka/CMSSW_14_0_0_pre2/src/EventFilter/EcalRawToDigi/python/ecalDigis_cff.py(67): <module>
  <frozen importlib._bootstrap>(228): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(850): exec_module
  <frozen importlib._bootstrap>(695): _load_unlocked
  <frozen importlib._bootstrap>(986): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1007): _find_and_load
  /data/user/treis/ecal/run3/dev/alpaka/CMSSW_14_0_0_pre2/src/Configuration/StandardSequences/python/RawToDigi_cff.py(13): <module>
  <frozen importlib._bootstrap>(228): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(850): exec_module
  <frozen importlib._bootstrap>(695): _load_unlocked
  <frozen importlib._bootstrap>(986): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1007): _find_and_load
  /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_0_pre2/src/FWCore/ParameterSet/python/Config.py(761): load
  step3_RAW2DIGI_RECO_VALIDATION_DQM.py(21): <module>

----- End Fatal Exception -------------------------------------------------
```
Either removing the `cuda` branch completely, or setting `cuda = _ecalDigisFromPortableProducer.clone()` works.

Reply:
> Trying this option I get

Ah, darn. I was expecting `clone()` and `toModify()` behaviors to be more similar, but see now that `SwitchProducer.clone()` treats `None` in an "ignoring" way, whereas the usual `clone()` and `toModify()` in a way that leads to the error. Sorry for the noise.

Reply:
> However, if the `gpu` and `alpaka` modifiers would be used together the former would have created a `cuda` branch that the later then does not modify. I guess they should never be used together though.

If the `gpu` and `alpaka` modifier are used at the same time, I would expect
  - the "cpu" branch to be replaced by the Alpaka version of the code
  - the "cuda" branch to be unaffected, and run the CUDA version of the code

Dos it work if you do
```python
_ecalDigisFromPortable = _ecalDigisFromPortableProducer.clone(
    digisInLabelEB = 'ecalDigisPortable:ebDigis',
    digisInLabelEE = 'ecalDigisPortable:eeDigis',
    produceDummyIntegrityCollections = True
)
alpaka.toModify(ecalDigis,
    cpu = _ecalDigisFromPortable.clone()
)
```
or more simply
```python
alpaka.toModify(ecalDigis,
    cpu = _ecalDigisFromPortableProducer.clone(
        digisInLabelEB = 'ecalDigisPortable:ebDigis',
        digisInLabelEE = 'ecalDigisPortable:eeDigis',
        produceDummyIntegrityCollections = True
    )
)
```
?

Reply:
Yes removing the `cuda` branch fro the `alpaka` modifier works.